skipruntime-ts/helpers/src/utils.ts

```diff
@@ -68,3 +71,96 @@ export function reactiveResponseHeader(
 ): [string, string] {
   return ["Skip-Reactive-Response-Token", JSON.stringify(reactiveResponse)];
 }
+
+/*****************************************************************************/
+// Joins
+/*****************************************************************************/
+
+type JoinObject = { value: JsonObject; side: "left" | "right" };
+
+class AddJoinSideField implements Mapper<Json, Json, Json, JoinObject> {
+  constructor(private joinSide: "left" | "right") {}
+
+  mapEntry(
+    key: Json,
+    values: NonEmptyIterator<JsonObject>,
+  ): Iterable<[Json, JoinObject]> {
+    const result: [Json, JoinObject][] = [];
+    for (const v of values) {
+      result.push([key, { value: v, side: this.joinSide }]);
+    }
+    return result;
+  }
+}
+
+function mergeObjects(object1: JoinObject, object2: JoinObject): Json {
+  const v1 = object1.value;
+  const v2 = object2.value;
+  const keys1 = Object.keys(v1);
+  if (keys1.some((key) => key in v2)) {
+    throw new Error(
+      "Objects don't have distinct fields: " +
+        JSON.stringify(v1) +
+        " " +
+        JSON.stringify(v2),
+    );
+  }
+  return { ...v1, ...v2 };
+}
+
+class MergeJoinFields implements Mapper<JoinObject, JoinObject, Json, Json> {
+  constructor(private allowMultipleValuesOnBothSides: boolean) {}
+
+  mapEntry(
+    key: Json,
+    values: NonEmptyIterator<JoinObject>,
+  ): Iterable<[Json, Json]> {
+    const result: [Json, Json][] = [];
+    let countLeft = 0;
+    let countRight = 0;
+    for (const value of values) {
+      if (value.side === "left") {
+        countLeft++;
+      }
+      if (value.side === "right") {
+        countRight++;
+      }
+    }
+    if (countLeft > 1 && countRight > 1) {
+      if (!this.allowMultipleValuesOnBothSides) {
+        throw new Error(
+          "More than one value detected on both sides for key: " +
+            JSON.stringify(key),
+        );
+      }
+    }
+    for (const value1 of values) {
+      if (value1.side === "left") {
+        for (const value2 of values) {
+          if (value2.side === "right") {
+            result.push([key, mergeObjects(value1, value2)]);
+          }
+        }
+      }
+    }
+    return result;
+  }
+}
+
+export function joinCollections<
```

Comment:
From an implementation point of view, I understand why this function is here. But from the API point of view, I think it would be better if this was a `join` operation in the `EagerCollection` interface. Also, if/when we reimplement this in skip, putting it in helpers will not be the right place anymore.

Reply:
You are right. I wish there was a clean way of extending a class in another file. You know something like:
```
extension class EagerCollection {
  join(...): ...
}
```